### PR TITLE
Handle the fact that noResolveResolution resolution is reused

### LINKED ISSUE #39795: TS Server fatal error:  Debug Failure. False expression.
Issue Type: <b>Bug</b>

❗️❗️❗️ Please fill in the sections below to help us diagnose the issue ❗️❗️❗️

**TypeScript Version:** 4.0.0-dev.20200727

**TS Server Log**
Server logging disabled. To help us fix crashes like this, please enable logging by setting:
```json
"typescript.tsserver.log": "verbose"
```
After enabling this setting, future crash reports will include the server log.

**TS Server Error Stack**
```
Error: Debug Failure. False expression.
    at watchFailedLookupLocationsOfExternalModuleResolutions (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:107070:26)
    at resolveNamesWithLocalCache (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:106932:21)
    at Object.resolveModuleNames (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:106986:20)
    at InferredProject.Project.resolveModuleNames (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:147629:45)
    at actualResolveModuleNamesWorker (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:102605:133)
    at resolveModuleNamesWorker (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:102828:26)
    at resolveModuleNamesReusingOldState (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:102902:24)
    at processImportedModules (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:104401:35)
    at findSourceFile (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:104137:17)
    at /Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:103966:85
    at getSourceFileFromReferenceWorker (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:103933:34)
    at processSourceFile (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:103966:13)
    at processRootFile (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:103789:13)
    at /Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:102706:60
    at Object.forEach (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:188:30)
    at Object.createProgram (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:102706:16)
    at synchronizeHostData (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:143253:26)
    at Proxy.getProgram (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:143341:13)
    at InferredProject.Project.updateGraphWorker (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:148163:53)
    at InferredProject.Project.updateGraph (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:148096:42)
    at ProjectService.assignOrphanScriptInfoToInferredProject (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:150164:25)
    at ProjectService.assignProjectToOpenedScriptInfo (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:151693:26)
    at /Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:151954:97
    at Object.flatMap (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:513:25)
    at ProjectService.applyChangesInOpenFiles (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:151954:41)
    at Session.handlers.ts.Map.ts.getEntries._a.<computed> (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:152947:46)
    at /Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:154805:88
    at IOSession.Session.executeWithRequestId (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:154796:28)
    at IOSession.Session.executeCommand (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:154805:33)
    at IOSession.Session.onMessage (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:154829:35)
    at Interface.<anonymous> (/Users/riginoommen/.vscode/extensions/ms-vscode.vscode-typescript-next-4.0.20200727/node_modules/typescript/lib/tsserver.js:157044:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
```

VS Code version: Code 1.47.2 (17299e413d5590b14ab0340ea477cdd86ff13daf, 2020-07-15T18:18:50.054Z)
OS version: Darwin x64 19.5.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 x 2800)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|5, 6, 9|
|Memory (System)|16.00GB (0.64GB free)|
|Process Argv|-psn_0_843982|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (14)</summary>

Extension|Author (truncated)|Version
---|---|---
html-snippets|abu|0.2.1
ng-template|Ang|0.1000.4
vscode-apollo|apo|1.16.6
gitlens|eam|10.2.2
env-switcher|Eck|0.2.0
vscode-html-css|ecm|0.2.3
beautify|Hoo|1.5.0
Angular2|joh|9.1.2
dotenv|mik|1.0.1
vscode-docker|ms-|1.4.1
vscode-typescript-next|ms-|4.0.20200727
vscode-graphql|Pri|0.2.14
html-preview-vscode|tht|0.2.5
vscode-wakatime|Wak|4.0.0

</details>

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #39795

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.0

File: src/compiler/resolutionCache.ts [hunk 1]
```diff
@@ -561,14 +561,15 @@ namespace ts {
...
+            noResolveResolution: T
         ) {
             if (resolution.refCount) {
                 resolution.refCount++;
                 Debug.assertDefined(resolution.files);
             }
             else {
                 resolution.refCount = 1;
-                Debug.assert(resolution.files === undefined);
+                Debug.assert(resolution.files === undefined || (resolution === noResolveResolution && length(noResolveResolution.files) === 0));
```
Comment (amcasey) [inline src/compiler/resolutionCache.ts:572]: Do `undefined` and `[]` have different meanings for `noResolveResolution.files`?  If not, why not just initialize it to `undefined` and create it the first time something is added?  (Mostly, it seems weird to me that `resolution` and `noResolveResolution` don't have the same shape.)

Comment (sheetalkamat) [reply src/compiler/resolutionCache.ts]: Instead of that i just changed the assert as we are really checking if no file is referencing this resolution.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.0 and LKG

**Pull request merged**